### ISSUE TITLE
Changing `bool` to `Result<(), ()>` in function return type leads to clippy warnings

### ISSUE THREAD

**Issue by tbu-** [rust-lang/rust-clippy#12890]
### Summary

I often see functions returning a success/failure via a `bool` in Rust code bases. Changing that to `Result<(), ()>`, while IMO objectively better than `bool`, since the meaning of success/failure is clear, gets linted by clippy. This means that people are encouraged not to switch away from `bool`.

I do not have a suggestion on how to fix this problem.

### Lint Name

result_unit_err

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

**Comment by Alexendoo** [comment]
The suggestion to provide an error type rather than `()` seems reasonable to me, is there anything that should be done if it could be?

**Comment by tbu-** [comment]
The problem is that the warning discourages the programmer to migrate away from `bool` as a return type.

Perhaps we could allow `Result<(), ()>` as only indicating success or failure?

**Comment by Alexendoo** [comment]
It discourages `Result<(), ()>`  sure, but `bool` APIs can move to either `Result<(), SomeError>` or a custom two variant enum if error doesn't seem like the right term

**Comment by tbu-** [comment]
> It discourages `Result<(), ()>` sure, but `bool` APIs can move to either `Result<(), SomeError>` or a custom two variant enum if error doesn't seem like the right term

Yes, they can. But they likely won't, because the transition from `Result<(), ()>` to `Result<(), CustomErrUnitStruct>` is a lot harder, it's like +10 lines of code because you suddenly need to implement `Display` and `Error`. Whereas staying on `bool` is simpler.

What I'm saying is that the way the lint currently works, it encourages staying on `bool` as opposed to moving to `Result<(), ()>`.

**Comment by Alexendoo** [comment]
I understand and agree, but I think it's fine as is. I agree with the lint that `Result<(), ()>` shouldn't appear in a public API but if it's good enough over `bool` there's always `#[allow]`
